Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/CancellationRate.hs

```diff
@@ -88,20 +97,72 @@ getCancellationRateData ::
-      then do
-        cancelledCount <- getCancellationCount windowSize driverId
-        let cancellationRate = (cancelledCount * 100) `div` max 1 assignedCount
-        pure (Just $ fromInteger assignedCount, Just $ fromInteger cancelledCount, Just $ fromInteger cancellationRate, merchantConfig.cancellationRateWindow)
-      else pure (Nothing, Nothing, Nothing, Nothing)
-  pure $ CancellationRateData assignedCount cancelledCount cancellationRate windowSize
+  let windowSize = findWindowSize merchantConfig
+  assignedCount <- getAssignedCount windowSize driverId
+  if (isJust merchantConfig.cancellationRateWindow) && (assignedCount > minimumRides)
+    then do
+      cancelledCount <- getCancellationCount windowSize driverId
+      let cancellationRate = (cancelledCount * 100) `div` max 1 assignedCount
+      pure $
+        Just $
+          CancellationRateData
+            { assignedCount = fromInteger assignedCount,
+              cancelledCount = fromInteger cancelledCount,
+              cancellationRate = fromInteger cancellationRate,
+              windowSize = fromMaybe 7 merchantConfig.cancellationRateWindow
+            }
+    else pure Nothing
   where
     findWindowSize merchantConfig = toInteger $ fromMaybe 7 merchantConfig.cancellationRateWindow
     findMinimumRides merchantConfig = toInteger $ fromMaybe 5 merchantConfig.cancellationRateCalculationThreshold
+
+nudgeOrBlockDriver ::
+  (MonadFlow m, CacheFlow m r, EsqDBFlow m r, CoreMetrics m, HasFlowEnv m r '["ltsCfg" ::: LocationTrackingeServiceConfig]) =>
+  TransporterConfig ->
+  Id DP.Person ->
+  m ()
+nudgeOrBlockDriver transporterConfig driverId = do
+  case (transporterConfig.cancellationRateWindow, transporterConfig.cancellationRateBasedNudgingAndBlockingConfig) of
+    (Just windowSize, Just CancellationRateBasedNudgingAndBlockingConfig {..}) -> do
+      driver <- QPerson.findById driverId >>= fromMaybeM (PersonNotFound driverId.getId)
+      (dailyCancellationRate, dailyAssignedCount) <- getCancellationRateOfDays 1 windowSize
+      (weeklyCancellationRate, weeklyAssignedCount) <- getCancellationRateOfDays 7 windowSize
+      blockedOnWeekly <- blockDriverCondition cancellationRateThresholdWeekly weeklyMinRidesforBlocking weeklyCancellationRate weeklyAssignedCount weeklyOffenceSuspensionTimeHours driver
+      unless blockedOnWeekly $ do
+        blockedOnDaily <- blockDriverCondition cancellationRateThresholdDaily dailyMinRidesforBlocking dailyCancellationRate dailyAssignedCount dailyOffenceSuspensionTimeHours driver
+        unless blockedOnDaily $ do
+          nudgeDriverCondition cancellationRateThresholdWeekly weeklyMinRidesforNudging weeklyMinRidesforBlocking weeklyCancellationRate weeklyAssignedCount driver FCM.CANCELLATION_RATE_NUDGE_WEEKLY
+          nudgeDriverCondition cancellationRateThresholdDaily dailyMinRidesforNudging dailyMinRidesforBlocking dailyCancellationRate dailyAssignedCount driver FCM.CANCELLATION_RATE_NUDGE_DAILY
```

Comment:
do we have to send these nudges everytime until driver is blocked or his cancellation rate is high?

Reply:
It's based on condition, inside the nudgeDriverCondition function you'll find a condition for it, only then we send. and this function is triggered on driver cancellation.

Reply:
cool